File: atcoder.jp/Rust (1.42.0)/abc149/abc149_d.rs
use proconio::{fastout, input};

macro_rules! max {
    ($x:expr) => ( $x );
    ($x:expr, $($xs:expr),+) => {
        std::cmp::max($x, max!( $($xs),+ ))
    };
}

#[fastout]
fn main() {
    input! {
        (n, k): (usize, usize),
        rsp: [usize; 3],
        t: String,
    }

    let t = t
        .chars()
        .map(|c| match c {
            'r' => 0,
            's' => 1,
            'p' => 2,
            _ => unreachable!(),
        })
        .collect::<Vec<_>>();

    let mut ans = 0;

    for i in 0..k {
        let iter = (i..n).step_by(k).collect::<Vec<_>>();
        let mut dp = vec![vec![0; 3]; iter.len()];
        match t[i] {
            0 => dp[0][2] = rsp[2],
            1 => dp[0][0] = rsp[0],
            2 => dp[0][1] = rsp[1],
            _ => unreachable!(),
        }
        for j in 1..iter.len() {
            let i = iter[j];
            for k in 0..3 {
                for l in 0..3 {
                    if k == l {
                        continue;
                    }
                    let score = if (t[i] - k as i32 + 3) % 3 == 1 {
                        rsp[k]
                    } else {
                        0
                    };
                    dp[j][k] = max!(dp[j][k], dp[j - 1][l] + score);
                }
            }
        }
        ans += dp.last().unwrap().iter().max().unwrap();
    }

    println!("{}", ans);
}

File: atcoder.jp/Python (CPython 3.11.4)/abc149/abc149_d.py
N, K = map(int, input().split())
SCORE = list(map(int, input().split()))
T = list(map(lambda s: 0 if s == "r" else 1 if s == "s" else 2, list(input())))

dp = [[-1] * 3 for _ in range(N)]

for k in range(K):
    for i in range(3):
        if (T[k] - i) % 3 == 1:
            score = SCORE[i]
        else:
            score = 0
        dp[k][i] = score
    for k in range(k + K, N, K):
        for i in range(3):
            for j in range(3):
                if j == i:
                    continue
                if (T[k] - i) % 3 == 1:
                    score = SCORE[i]
                else:
                    score = 0
                dp[k][i] = max(dp[k][i], dp[k - K][j] + score)

ans = 0
for i in range(K):
    ans += max(dp[N - i - 1])

print(ans)
